### ISSUE TITLE
표준 ML 개발 프로젝트 패턴

### ISSUE THREAD

**Issue by Sarang-Han** [Sarang-Han/2048-rl-project#2]
## 프로젝트 구조 예시 (프로덕션 지향)

```
ml-project/
├── data/                # (로컬 샘플 데이터, DVC/S3 같은 외부 스토리지 연동)
│   ├── raw/             # 원천 데이터
│   ├── processed/       # 전처리된 데이터
│   └── external/        # 외부 데이터셋
│
├── notebooks/           # Jupyter 실험용 (EDA, 모델 탐색)
│
├── src/                 # 실제 라이브러리 코드
│   ├── data/            # 데이터 로딩/전처리 스크립트
│   ├── features/        # feature engineering
│   ├── models/          # 모델 정의 (PyTorch, TensorFlow 등)
│   ├── training/        # 학습 루프, trainer 모듈
│   ├── evaluation/      # metrics, evaluation 함수
│   └── utils/           # 공통 유틸
│
├── configs/             # Hydra/OmegaConf 기반 설정 (yaml)
│   ├── model/
│   ├── data/
│   └── train/
│
├── experiments/         # 결과 저장 (로그, 체크포인트, metrics)
│
├── scripts/             # 실행 스크립트 (bash, CLI entrypoints)
│   ├── train.py
│   ├── evaluate.py
│   └── predict.py
│
├── tests/               # pytest 기반 유닛테스트
│
├── requirements.txt     # 의존성 (또는 pyproject.toml/poetry.lock)
├── Dockerfile           # 컨테이너화
├── dvc.yaml             # (옵션) 데이터/모델 파이프라인 정의
├── .gitignore
└── README.md
```

**Comment by Sarang-Han** [comment]
## 기술스택 예시

**언어 & 프레임워크**

- Python 3.10+ (표준)
- PyTorch / TensorFlow
- JAX (최신 연구나 구글식 스택에서 인기)

**데이터 관리**

- pandas / Polars (ETL)
- PyArrow / Parquet (대규모 데이터 저장)
- DVC / MLflow / Weights & Biases (W&B) → 데이터 버저닝 + 실험 관리

**환경 관리**

- Poetry / Pipenv (또는 conda, 하지만 최근엔 poetry 선호 ↑)
- Docker (재현성 확보)

**설정/구성 관리**

- Hydra / OmegaConf (yaml 기반)
- dotenv (환경 변수)

**로깅 & 실험 추적**

- TensorBoard (baseline)
- W&B / MLflow

**테스트 & 품질 관리**

- pytest
- pre-commit hooks (black, isort, flake8, mypy, ruff)

**MLOps**

- CI/CD: GitHub Actions, GitLab CI
- 모델 서빙: FastAPI / BentoML / TorchServe
- 배포: Docker + Kubernetes (k8s), AWS Sagemaker, GCP Vertex AI, Azure ML

## 참고

https://github.com/michaelriedl/2048-ppo-agent

**Comment by Sarang-Han** [comment]
## EDA란

- 목적: 데이터를 이해하고, 가설을 세우기 위한 탐색 과정

- 주로 하는 것:
      - **결측치 비율 확인**
      - **분포 시각화 (히스토그램, 박스플롯, kde)**
      - **이상치 탐색**
      - **변수 간 상관관계 (heatmap, scatterplot)**
      - **타겟(label) 불균형 확인**

- 도구: `pandas`, `matplotlib`, `seaborn`, `plotly`


## EDA 기본 패턴

1. **데이터 크기/타입 확인** - `df.shape`, `df.info()`, `df.head()`

2. **결측치 탐색** - `df.isnull().sum()`

3. **기초 통계** - `df.describe()`

4. **타겟 분포 확인**
    - 분류 → 클래스 불균형 여부
    - 회귀 → 분포 치우침 여부

5. **피처별 분포 확인**
   - 수치형: 히스토그램/박스플롯
   - 범주형: `value_counts`, `barplot`

6. **피처 간 관계 확인**
   - 수치형: 상관계수 heatmap
   - 범주형 vs 타겟: 카이제곱, boxplot

7. **이상치 탐색**
   - IQR, z-score
   - 시각화로 outlier 패턴 확인

8. **특정 도메인 룰 기반 검사**
   - 예: 시간 데이터 → 누락된 날짜, 주기성
   - 텍스트 → 토큰 분포, 단어 빈도

프로젝트에서 위치는 주로 -> `notebooks/EDA.ipynb`
